### PR TITLE
API: add `From` field to JSONPatchOperation

### PR DESCRIPTION
**Pull request by soulxu**
**What type of PR is this?**
* "feat(translator): add new feature"

**What this PR does / why we need it**:
There is a requirement to use the PrivateKeyProvider in Envoy Gateway by using EnvoyPatchPolicy API. https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-privatekeyprovider

The move operation is needed in this case. The move operation was supported by jsonpatch https://datatracker.ietf.org/doc/html/rfc6902

But the move operation needs two parameter from is the `path`, another is the `from`. The `from` was exposed in the EnvoyPatchPolicy API. This PR add that new field.

**Which issue(s) this PR fixes**:
Related to issues: https://github.com/envoyproxy/gateway/issues/2456

### CODE REVIEW

**Comment by soulxu** [comment]
cc @arkodg

File: api/v1alpha1/envoypatchpolicy_types.go [hunk 1]
```diff
@@ -112,6 +112,10 @@ type JSONPatchOperation struct {
 	// Path is the location of the target document/field where the operation will be performed
 	// Refer to https://datatracker.ietf.org/doc/html/rfc6901 for more details.
 	Path string `json:"path"`
+	// From is the location of the source document/field where the operation will be performed
+	// Refer to https://datatracker.ietf.org/doc/html/rfc6901 for more details.
+	// +optional
+	From string `json:"from,omitempty"`
```
[the remark this answers is not in the document]
Comment (arkodg) [reply api/v1alpha1/envoypatchpolicy_types.go:118]: or does that break the implementation ?

**Comment by arkodg** [comment]
hey @soulxu does this PR include implementation as well ? if so, can you add a test case similar to https://github.com/envoyproxy/gateway/blob/main/internal/xds/translator/testdata/in/xds-ir/jsonpatch.yaml ?

File: api/v1alpha1/envoypatchpolicy_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (soulxu) [reply api/v1alpha1/envoypatchpolicy_types.go:118]: a ptr works here also. I only need to change the IR struct using the ptr also.

**Comment by soulxu** [comment]
> hey @soulxu does this PR include implementation as well ? if so, can you add a test case similar to https://github.com/envoyproxy/gateway/blob/main/internal/xds/translator/testdata/in/xds-ir/jsonpatch.yaml ?

yes, this includes the implementation, I will add a test for it.

File: api/v1alpha1/envoypatchpolicy_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (soulxu) [reply api/v1alpha1/envoypatchpolicy_types.go:118]: @arkodg may I ask the reason you prefer ptr at here? I also see other case in the API definition not use ptr when it is optional.

[the remark this answers is not in the document]
Comment (arkodg) [reply api/v1alpha1/envoypatchpolicy_types.go:118]: here's the recommendation from upstream https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required
ptr ensures field has `nil` type

[the remark this answers is not in the document]
Comment (soulxu) [reply api/v1alpha1/envoypatchpolicy_types.go:118]: @arkodg thanks for point me that!

**Comment by soulxu** [comment]
Sorry for the force push, I messed up with dco.

File: api/v1alpha1/envoypatchpolicy_types.go [hunk 2]
```diff
@@ -112,6 +112,10 @@ type JSONPatchOperation struct {
 	// Path is the location of the target document/field where the operation will be performed
 	// Refer to https://datatracker.ietf.org/doc/html/rfc6901 for more details.
 	Path string `json:"path"`
+	// From is the location of the source document/field where the operation will be performed
```
Comment (zhaohuabing) [inline api/v1alpha1/envoypatchpolicy_types.go:115]:
```suggestion
	// From is the source location of the value to be copied or moved
	// Only valid for move or copy operations
```

Comment (zhaohuabing) [reply api/v1alpha1/envoypatchpolicy_types.go]: We probably don't have use cases for copy :-)

File: internal/xds/translator/testdata/out/xds-ir/jsonpatch.endpoints.yaml [hunk 3]
```diff
@@ -7,6 +7,7 @@
             address: 1.2.3.4
             portValue: 50000
       loadBalancingWeight: 1
+    loadBalancingWeight: 50
```
Comment (zirain) [inline internal/xds/translator/testdata/out/xds-ir/jsonpatch.endpoints.yaml:10]: is this valid?

Comment (soulxu) [reply internal/xds/translator/testdata/out/xds-ir/jsonpatch.endpoints.yaml:10]: Yes, I think it is https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/endpoint/v3/endpoint_components.proto#envoy-v3-api-msg-config-endpoint-v3-localitylbendpoints

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zhaohuabing — APPROVED** [review]
LGTM. Thanks for the improvement!

**Pull request merged**